File: seqspec/seqspec_file.py
"""File module for seqspec.

This module provides functionality to list and format files present in seqspec files.
"""

import json
from argparse import ArgumentParser, Namespace, RawTextHelpFormatter
from collections import defaultdict
from pathlib import Path
from typing import Dict, List, Optional

from seqspec import seqspec_find
from seqspec.Assay import Assay
from seqspec.File import File
from seqspec.utils import load_spec


def setup_file_args(parser) -> ArgumentParser:
    """Create and configure the file command subparser."""
    subparser = parser.add_parser(
        "file",
        description="""
List files present in seqspec file.

Examples:
seqspec file -m rna spec.yaml                                          # List paired read files
seqspec file -m rna -f interleaved spec.yaml                           # List interleaved read files
seqspec file -m rna -f list -k url spec.yaml                           # List urls of all read files
seqspec file -m rna -f list -s region -k all spec.yaml                 # List all files in regions
seqspec file -m rna -f json -s region-type -k all -i barcode spec.yaml # List files for barcode regions in json
---
""",
        help="List files present in seqspec file",
        formatter_class=RawTextHelpFormatter,
    )
    subparser_required = subparser.add_argument_group("required arguments")

    subparser.add_argument("yaml", help="Sequencing specification yaml file", type=Path)
    subparser.add_argument(
        "-o",
        "--output",
        metavar="OUT",
        help="Path to output file",
        type=Path,
        default=None,
    )
    subparser.add_argument(
        "-i",
        "--ids",
        metavar="IDs",
        help="Ids to list",
        type=str,
        default=None,
    )
    subparser_required.add_argument(
        "-m",
        "--modality",
        metavar="MODALITY",
        help="Modality",
        type=str,
        required=True,
    )
    choices = ["read", "region", "file", "region-type"]
    subparser.add_argument(
        "-s",
        "--selector",
        metavar="SELECTOR",
        help=f"Selector for ID, [{', '.join(choices)}] (default: read)",
        type=str,
        default="read",
        choices=choices,
    )
    choices = ["paired", "interleaved", "index", "list", "json"]
    subparser.add_argument(
        "-f",
        "--format",
        metavar="FORMAT",
        help=f"Format, [{', '.join(choices)}], default: paired",
        type=str,
        default="paired",
        choices=choices,
    )
    choices = [
        "file_id",
        "filename",
        "filetype",
        "filesize",
        "url",
        "urltype",
        "md5",
        "all",
    ]
    subparser.add_argument(
        "-k",
        "--key",
        metavar="KEY",
        help=f"Key, [{', '.join(choices)}], default: file_id",
        type=str,
        default="file_id",
        choices=choices,
    )

    # option to get the full path of the file
    subparser.add_argument(
        "--fullpath",
        help="Use full path for local files",
        action="store_true",
        default=False,
    )

    return subparser


def validate_file_args(parser: ArgumentParser, args: Namespace) -> None:
    """Validate the file command arguments."""
    if not Path(args.yaml).exists():
        parser.error(f"Input file does not exist: {args.yaml}")

    if args.output and Path(args.output).exists() and not Path(args.output).is_file():
        parser.error(f"Output path exists but is not a file: {args.output}")

    if args.key in ["filesize", "filetype", "urltype", "md5"] and args.format in [
        "paired",
        "interleaved",
        "index",
    ]:
        parser.error(
            f"Format '{args.format}' valid only with key 'file_id', 'filename', or 'url'"
        )


def seqspec_file(
    spec: Assay,
    modality: str,
    ids: Optional[List[str]] = None,
    selector: str = "read",
) -> Dict[str, List[File]]:
    """Core functionality to list files from a seqspec.

    Args:
        spec: The Assay object to operate on
        spec_fn: Path to the spec file, used for relative path resolution
        modality: The modality to list files for
        ids: Optional list of IDs to filter by
        selector: Type of ID to filter by (read, region, file, region-type)

    Returns:
        Dictionary mapping IDs to lists of File objects
    """
    # NOTE: LIST FILES DOES NOT RESPECT ORDERING OF INPUT IDs LIST
    # NOTE: seqspec file -s read gets the files for the read, not the files mapped from the regions associated with the read.
    LIST_FILES = {
        "read": list_read_files,
        "region": list_region_files,
        "file": list_all_files,
    }

    LIST_FILES_BY_ID = {
        "read": list_files_by_read_id,
        "file": list_files_by_file_id,
        "region": list_files_by_region_id,
        "region-type": list_files_by_region_type,
    }

    # Get files based on whether we're filtering by IDs
    if not ids:
        # list all files
        return LIST_FILES[selector](spec, modality)
    else:
        # list files by id
        return LIST_FILES_BY_ID[selector](spec, modality, ids)


def run_file(parser: ArgumentParser, args: Namespace) -> None:
    """Run the file command."""
    validate_file_args(parser, args)

    spec = load_spec(args.yaml)
    ids = args.ids.split(",") if args.ids else []

    files = seqspec_file(
        spec=spec,
        modality=args.modality,
        ids=ids,
        selector=args.selector,
    )

    if files:
        FORMAT = {
            "list": format_list_files_metadata,
            "paired": format_list_files,
            "interleaved": format_list_files,
            "index": format_list_files,
            "json": format_json_files,
        }

        result = FORMAT[args.format](
            files, args.format, args.key, Path(args.yaml), args.fullpath
        )

        if args.output:
            args.output.write_text(str(result))
        else:
            print(result)


def list_read_files(spec: Assay, modality: str) -> Dict[str, List[File]]:
    """List files for all reads in a modality."""
    files = defaultdict(list)
    reads = spec.get_seqspec(modality)
    for rd in reads:
        if rd.files:
            files[rd.read_id] = rd.files
    return files


def list_all_files(spec: Assay, modality: str) -> Dict[str, List[File]]:
    """List all files in a modality."""
    files_rd = list_read_files(spec, modality)
    files_rgn = list_region_files(spec, modality)
    return {**files_rd, **files_rgn}


def list_onlist_files(spec, modality):
    files = defaultdict(list)
    regions = spec.get_libspec(modality).get_onlist_regions()
    for r in regions:
        if r.onlist is None:
            continue
        files[r.region_id].append(r.onlist)
    return files


def list_region_files(spec, modality):
    return list_onlist_files(spec, modality)


def format_list_files_metadata(
    files: Dict[str, List[File]],
    fmt: str,
    k: str,
    spec_fn: Path = Path(""),
    fp: bool = False,
) -> str:
    """Format file metadata as a tab-separated list."""
    x = []
    if k == "all":
        for items in zip(*files.values()):
            for key, item in zip(files.keys(), items):
                x.append(
                    f"{key}\t{item.file_id}\t{item.filename}\t{item.filetype}\t{item.filesize}\t{item.url}\t{item.urltype}\t{item.md5}"
                )
    else:
        for items in zip(*files.values()):
            for key, item in zip(files.keys(), items):
                attr = str(getattr(item, k))
                id = item.file_id
                x.append(f"{key}\t{id}\t{attr}")
    return "\n".join(x)


def format_json_files(
    files: Dict[str, List[File]],
    fmt: str,
    k: str,
    spec_fn: Path = Path(""),
    fp: bool = False,
) -> str:
    """Format files as JSON."""
    x = []
    for items in zip(*files.values()):
        if k == "all":
            for key, item in zip(files.keys(), items):
                d = item.to_dict()
                if item.urltype == "local" and fp:
                    d["url"] = str(spec_fn.parent / d["url"])
                x.append(d)
        else:
            for key, item in zip(files.keys(), items):
                attr = getattr(item, k)
                if k == "url" and item.urltype == "local" and fp:
                    attr = str(spec_fn.parent / attr)
                x.append({"file_id": item.file_id, k: attr})
    return json.dumps(x, indent=4)


def format_list_files(
    files: Dict[str, List[File]],
    fmt: str,
    k: Optional[str] = None,
    spec_fn: Path = Path(""),
    fp: bool = False,
) -> str:
    """Format files as a list based on the format type."""
    x = []

    if fmt == "paired":
        for items in zip(*files.values()):
            t = []
            for i in items:
                if k:
                    attr = str(getattr(i, k))
                    if k == "url" and i.urltype == "local" and fp:
                        attr = str(spec_fn.parent / attr)
                    t.append(attr)
                else:
                    t.append(i.filename)
            x.append("\t".join(t))

    elif fmt in ["interleaved", "list"]:
        for items in zip(*files.values()):
            for item in items:
                id = item.filename
                if k:
                    id = str(getattr(item, k))
                    if k == "url" and item.urltype == "local" and fp:
                        id = str(spec_fn.parent / id)
                x.append(id)

    elif fmt == "index":
        t = []
        for items in zip(*files.values()):
            for item in items:
                id = item.filename
                if k:
                    id = str(getattr(item, k))
                    if k == "url" and item.urltype == "local" and fp:
                        id = str(spec_fn.parent / id)
                t.append(id)
        x.append(",".join(t))

    return "\n".join(x)


def list_files_by_read_id(
    spec: Assay, modality: str, read_ids: List[str]
) -> Dict[str, List[File]]:
    """List files for specific read IDs."""
    seqspec = spec.get_seqspec(modality)
    files = defaultdict(list)
    ids = set(read_ids)
    # TODO return the files in the order of the ids given in the input
    # NOTE ORDERING HERE IS IMPORANT SEE GET_INDEX_BY_FILES FUNCTION
    for read in seqspec:
        if read.read_id in ids and read.files:
            files[read.read_id].extend(read.files)
    return files


def list_files_by_file_id(
    spec: Assay, modality: str, file_ids: List[str]
) -> Dict[str, List[File]]:
    """List files for specific file IDs."""
    seqspec = spec.get_seqspec(modality)
    ids = set(file_ids)
    files = defaultdict(list)
    # TODO: NOTE ORDERING HERE IS IMPORTANT SEE RUN_LIST_FILES FUNCTION
    for read in seqspec:
        if read.files:
            for file in read.files:
                if file.filename in ids:
                    files[read.read_id].append(file)
    return files


def list_files_by_region_id(
    spec: Assay, modality: str, region_ids: List[str]
) -> Dict[str, List[File]]:
    """List files for specific region IDs."""
    files = list_region_files(spec, modality)
    ids = set(region_ids)
    new_files = defaultdict(list)
    for region_id, region_files in files.items():
        if region_id in ids:
            new_files[region_id].extend(region_files)
    return new_files


def list_files_by_region_type(
    spec: Assay, modality: str, region_types: List[str]
) -> Dict[str, List[File]]:
    """List files for specific region types."""
    files = list_region_files(spec, modality)
    ids = set(region_types)
    new_files = defaultdict(list)
    for region_id, region_files in files.items():
        r = seqspec_find.find_by_region_id(spec, modality, region_id)[0]
        if r.region_type in ids:
            new_files[region_id].extend(region_files)
    return new_files

File: src/seqspec_file.rs
use crate::auth::RemoteAccess;
use crate::models::assay::Assay;
use crate::models::file::File;
use crate::models::onlist::Onlist;
use crate::utils;
use clap::Args;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};

#[derive(Debug, Args)]
pub struct FileArgs {
    #[clap(short, long, help = "Output file path", value_name = "OUT")]
    output: Option<PathBuf>,

    #[clap(help = "Path or URL to sequencing specification YAML", required = true)]
    yaml: String,

    #[clap(
        short,
        long,
        help = "IDs",
        value_name = "IDs",
        required = false,
        value_delimiter = ','
    )]
    ids: Option<Vec<String>>,

    #[clap(
        short,
        long,
        help = "Modality",
        value_name = "MODALITY",
        required = true
    )]
    modality: String,

    #[clap(
        short,
        long,
        help = "Selector",
        value_name = "SELECTOR",
        value_parser = ["read", "region", "file", "region-type"],
        default_value = "read"
    )]
    selector: String,

    #[clap(
        short = 'f',
        long,
        help = "Format",
        value_name = "FORMAT",
        value_parser = ["paired", "interleaved", "index", "list", "json"],
        default_value = "paired"
    )]
    format: String,

    #[clap(
        short,
        long,
        help = "Key",
        value_name = "KEY",
        value_parser = ["file_id", "filename", "filetype", "filesize", "url", "urltype", "md5", "all"],
        default_value = "file_id"
    )]
    key: String,

    #[clap(long, help = "Use full path for local urls", default_value = "false")]
    fullpath: bool,

    #[clap(long, env = "SEQSPEC_AUTH_PROFILE", value_name = "PROFILE")]
    auth_profile: Option<String>,
}

pub fn run_file(args: &FileArgs) {
    let remote_access = RemoteAccess::load(args.auth_profile.as_deref()).unwrap_or_else(|err| {
        eprintln!("{}", err);
        std::process::exit(1);
    });
    validate_file_args(args, &remote_access);
    let spec = utils::load_spec_source(&args.yaml, &remote_access).unwrap_or_else(|err| {
        eprintln!("{}", err);
        std::process::exit(1);
    });
    let spec_base = utils::spec_base_from_source(&args.yaml);

    let ids = args.ids.clone();
    let files = seqspec_file(&spec, &args.modality, ids.as_ref(), &args.selector);

    if !files.is_empty() {
        let result = match args.format.as_str() {
            "list" => format_list_files_metadata(&files, &args.key, spec_base.as_deref(), args.fullpath),
            "paired" | "interleaved" | "index" => format_list_files(
                &files,
                &args.format,
                Some(&args.key),
                spec_base.as_deref(),
                args.fullpath,
            ),
            "json" => format_json_files(&files, &args.key, spec_base.as_deref(), args.fullpath),
            _ => String::new(),
        };

        if let Some(output) = &args.output {
            let mut file = fs::File::create(output).unwrap();
            writeln!(file, "{}", result).unwrap();
        } else {
            println!("{}", result);
        }
    }
}

fn validate_file_args(args: &FileArgs, remote_access: &RemoteAccess) {
    if let Err(err) = utils::validate_source_exists(&args.yaml, remote_access) {
        eprintln!("{}", err);
        std::process::exit(1);
    }
    if ["filesize", "filetype", "urltype", "md5"].contains(&args.key.as_str())
        && ["paired", "interleaved", "index"].contains(&args.format.as_str())
    {
        eprintln!(
            "Format '{}' valid only with key 'file_id', 'filename', or 'url'",
            args.format
        );
        std::process::exit(1);
    }
}

pub fn seqspec_file(
    spec: &Assay,
    modality: &String,
    ids: Option<&Vec<String>>,
    selector: &String,
) -> HashMap<String, Vec<File>> {
    let list_files = |sel: &str| -> HashMap<String, Vec<File>> {
        match sel {
            "read" => list_read_files(spec, modality),
            "region" => list_region_files(spec, modality),
            "file" => list_all_files(spec, modality),
            _ => HashMap::new(),
        }
    };

    let list_files_by_id = |sel: &str, ids: &Vec<String>| -> HashMap<String, Vec<File>> {
        match sel {
            "read" => list_files_by_read_id(spec, modality, ids),
            "file" => list_files_by_file_id(spec, modality, ids),
            "region" => list_files_by_region_id(spec, modality, ids),
            "region-type" => list_files_by_region_type(spec, modality, ids),
            _ => HashMap::new(),
        }
    };

    match ids {
        None => list_files(&selector),
        Some(v) if v.is_empty() => list_files(&selector),
        Some(v) => list_files_by_id(&selector, v),
    }
}

fn list_read_files(spec: &Assay, modality: &String) -> HashMap<String, Vec<File>> {
    let mut files: HashMap<String, Vec<File>> = HashMap::new();
    for rd in spec.get_seqspec(modality) {
        if !rd.files.is_empty() {
            files.insert(rd.read_id, rd.files);
        }
    }
    files
}

fn list_all_files(spec: &Assay, modality: &String) -> HashMap<String, Vec<File>> {
    let mut rd = list_read_files(spec, modality);
    let rgn = list_region_files(spec, modality);
    rd.extend(rgn);
    rd
}

fn list_onlist_files(spec: &Assay, modality: &String) -> HashMap<String, Vec<Onlist>> {
    let mut files: HashMap<String, Vec<Onlist>> = HashMap::new();
    let regions = spec.get_libspec(modality).unwrap().get_onlist_regions();
    for r in regions {
        if let Some(ol) = r.onlist {
            files.entry(r.region_id).or_default().push(ol);
        }
    }
    files
}

fn list_region_files(spec: &Assay, modality: &String) -> HashMap<String, Vec<File>> {
    // convert Onlist to File-shaped map by copying fields
    let onlists = list_onlist_files(spec, modality);
    let mut files: HashMap<String, Vec<File>> = HashMap::new();
    for (region_id, v) in onlists {
        let mut out: Vec<File> = Vec::new();
        for ol in v {
            out.push(File {
                file_id: ol.file_id,
                filename: ol.filename,
                filetype: ol.filetype,
                filesize: ol.filesize,
                url: ol.url,
                urltype: ol.urltype,
                md5: ol.md5,
            });
        }
        files.insert(region_id, out);
    }
    files
}

fn format_list_files_metadata(
    files: &HashMap<String, Vec<File>>,
    k: &String,
    _spec_base: Option<&Path>,
    _fp: bool,
) -> String {
    let mut x: Vec<String> = Vec::new();
    for row in iter_file_rows(files) {
        if k == "all" {
            for (key, item) in row {
                x.push(format!(
                    "{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}",
                    key,
                    item.file_id,
                    item.filename,
                    item.filetype,
                    item.filesize,
                    item.url,
                    item.urltype,
                    item.md5
                ));
            }
        } else {
            for (key, item) in row {
                let attr = match k.as_str() {
                    "url" => item.url.clone(),
                    "file_id" => item.file_id.clone(),
                    "filename" => item.filename.clone(),
                    "filetype" => item.filetype.clone(),
                    "filesize" => item.filesize.to_string(),
                    "urltype" => item.urltype.clone(),
                    "md5" => item.md5.clone(),
                    _ => String::new(),
                };
                x.push(format!("{}\t{}\t{}", key, item.file_id, attr));
            }
        }
    }
    x.join("\n")
}

fn format_json_files(
    files: &HashMap<String, Vec<File>>,
    k: &String,
    spec_base: Option<&Path>,
    fp: bool,
) -> String {
    use serde_json::json;
    let mut x: Vec<serde_json::Value> = Vec::new();
    for row in iter_file_rows(files) {
        for (_key, item) in row {
            if k == "all" {
                let mut d = serde_json::to_value(item).unwrap();
                if item.urltype == "local" && fp {
                    if let Some(obj) = d.as_object_mut() {
                        obj.insert("url".to_string(), json!(full_path(spec_base, &item.url)));
                    }
                }
                x.push(d);
            } else {
                let mut attr = match k.as_str() {
                    "url" => maybe_full(&item.url, &item.urltype, spec_base, fp),
                    _ => String::new(),
                };
                if k != "url" {
                    attr = match k.as_str() {
                        "file_id" => item.file_id.clone(),
                        "filename" => item.filename.clone(),
                        "filetype" => item.filetype.clone(),
                        "filesize" => item.filesize.to_string(),
                        "urltype" => item.urltype.clone(),
                        "md5" => item.md5.clone(),
                        _ => String::new(),
                    };
                }
                x.push(json!({ "file_id": item.file_id, k: attr }));
            }
        }
    }
    serde_json::to_string_pretty(&x).unwrap()
}

fn format_list_files(
    files: &HashMap<String, Vec<File>>,
    fmt: &String,
    k: Option<&String>,
    spec_base: Option<&Path>,
    fp: bool,
) -> String {
    let mut out: Vec<String> = Vec::new();
    if fmt == "paired" {
        for row in iter_file_rows(files) {
            let mut t: Vec<String> = Vec::new();
            for (_key, i) in row {
                let val = if let Some(key) = k {
                    let mut attr = match key.as_str() {
                        "url" => maybe_full(&i.url, &i.urltype, spec_base, fp),
                        _ => String::new(),
                    };
                    if key != &"url".to_string() {
                        attr = match key.as_str() {
                            "file_id" => i.file_id.clone(),
                            "filename" => i.filename.clone(),
                            "filetype" => i.filetype.clone(),
                            "filesize" => i.filesize.to_string(),
                            "urltype" => i.urltype.clone(),
                            "md5" => i.md5.clone(),
                            _ => String::new(),
                        };
                    }
                    attr
                } else {
                    i.filename.clone()
                };
                t.push(val);
            }
            out.push(t.join("\t"));
        }
    } else if fmt == "interleaved" || fmt == "list" {
        for row in iter_file_rows(files) {
            for (_key, i) in row {
                let id = if let Some(key) = k {
                    let mut attr = match key.as_str() {
                        "url" => maybe_full(&i.url, &i.urltype, spec_base, fp),
                        _ => String::new(),
                    };
                    if key != &"url".to_string() {
                        attr = match key.as_str() {
                            "file_id" => i.file_id.clone(),
                            "filename" => i.filename.clone(),
                            "filetype" => i.filetype.clone(),
                            "filesize" => i.filesize.to_string(),
                            "urltype" => i.urltype.clone(),
                            "md5" => i.md5.clone(),
                            _ => String::new(),
                        };
                    }
                    attr
                } else {
                    i.filename.clone()
                };
                out.push(id);
            }
        }
    } else if fmt == "index" {
        let mut t: Vec<String> = Vec::new();
        for row in iter_file_rows(files) {
            for (_key, i) in row {
                let id = if let Some(key) = k {
                    let mut attr = match key.as_str() {
                        "url" => maybe_full(&i.url, &i.urltype, spec_base, fp),
                        _ => String::new(),
                    };
                    if key != &"url".to_string() {
                        attr = match key.as_str() {
                            "file_id" => i.file_id.clone(),
                            "filename" => i.filename.clone(),
                            "filetype" => i.filetype.clone(),
                            "filesize" => i.filesize.to_string(),
                            "urltype" => i.urltype.clone(),
                            "md5" => i.md5.clone(),
                            _ => String::new(),
                        };
                    }
                    attr
                } else {
                    i.filename.clone()
                };
                t.push(id);
            }
        }
        out.push(t.join(","));
    }
    out.join("\n")
}

fn ordered_file_columns<'a>(
    files: &'a HashMap<String, Vec<File>>,
) -> Vec<(&'a String, &'a Vec<File>)> {
    let mut columns: Vec<(&String, &Vec<File>)> = files.iter().collect();
    columns.sort_by(|(left, _), (right, _)| left.cmp(right));
    columns
}

fn iter_file_rows<'a>(files: &'a HashMap<String, Vec<File>>) -> Vec<Vec<(&'a String, &'a File)>> {
    let columns = ordered_file_columns(files);
    let row_count = columns
        .iter()
        .map(|(_, items)| items.len())
        .min()
        .unwrap_or(0);
    let mut rows = Vec::new();
    for row_idx in 0..row_count {
        let mut row = Vec::new();
        for (key, items) in &columns {
            row.push((*key, &items[row_idx]));
        }
        rows.push(row);
    }
    rows
}

fn list_files_by_read_id(
    spec: &Assay,
    modality: &String,
    read_ids: &Vec<String>,
) -> HashMap<String, Vec<File>> {
    let mut files: HashMap<String, Vec<File>> = HashMap::new();
    let ids: HashSet<String> = read_ids.iter().cloned().collect();
    for read in spec.get_seqspec(modality) {
        if ids.contains(&read.read_id) && !read.files.is_empty() {
            files.entry(read.read_id).or_default().extend(read.files);
        }
    }
    files
}

fn list_files_by_file_id(
    spec: &Assay,
    modality: &String,
    file_ids: &Vec<String>,
) -> HashMap<String, Vec<File>> {
    let mut files: HashMap<String, Vec<File>> = HashMap::new();
    let ids: HashSet<String> = file_ids.iter().cloned().collect();
    for read in spec.get_seqspec(modality) {
        for file in read.files {
            if ids.contains(&file.filename) {
                files.entry(read.read_id.clone()).or_default().push(file);
            }
        }
    }
    files
}

fn list_files_by_region_id(
    spec: &Assay,
    modality: &String,
    region_ids: &Vec<String>,
) -> HashMap<String, Vec<File>> {
    let files = list_region_files(spec, modality);
    let ids: HashSet<String> = region_ids.iter().cloned().collect();
    let mut new_files: HashMap<String, Vec<File>> = HashMap::new();
    for (region_id, region_files) in files {
        if ids.contains(&region_id) {
            new_files.entry(region_id).or_default().extend(region_files);
        }
    }
    new_files
}

fn list_files_by_region_type(
    spec: &Assay,
    modality: &String,
    region_types: &Vec<String>,
) -> HashMap<String, Vec<File>> {
    let files = list_region_files(spec, modality);
    let ids: HashSet<String> = region_types.iter().cloned().collect();
    let mut new_files: HashMap<String, Vec<File>> = HashMap::new();
    for (region_id, region_files) in files {
        let m = spec.get_libspec(modality).unwrap();
        let regions = m.get_region_by_id(&region_id);
        let r = regions.first().unwrap().clone();
        if ids.contains(&r.region_type) {
            new_files.entry(region_id).or_default().extend(region_files);
        }
    }
    new_files
}

fn maybe_full(url: &String, urltype: &String, spec_base: Option<&Path>, fp: bool) -> String {
    if urltype == "local" && fp {
        full_path(spec_base, url)
    } else {
        url.clone()
    }
}

fn full_path(spec_base: Option<&Path>, url: &String) -> String {
    spec_base
        .map(|base| base.join(url).to_string_lossy().to_string())
        .unwrap_or_else(|| url.clone())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::load_spec;

    fn dogma_spec() -> Assay {
        load_spec(&PathBuf::from("tests/fixtures/spec.yaml"))
    }

    #[test]
    fn test_seqspec_file_read_selector() {
        let spec = dogma_spec();
        let files = seqspec_file(&spec, &"rna".into(), None, &"read".into());
        assert!(!files.is_empty());
        // Each key should be a read_id
        for (read_id, file_list) in &files {
            assert!(!read_id.is_empty());
            assert!(!file_list.is_empty());
        }
    }

    #[test]
    fn test_seqspec_file_region_selector() {
        let spec = dogma_spec();
        let files = seqspec_file(&spec, &"rna".into(), None, &"region".into());
        // Region files come from onlists, which the rna modality should have
        assert!(!files.is_empty());
    }

    #[test]
    fn test_seqspec_file_all_selector() {
        let spec = dogma_spec();
        let all_files = seqspec_file(&spec, &"rna".into(), None, &"file".into());
        let read_files = seqspec_file(&spec, &"rna".into(), None, &"read".into());
        let region_files = seqspec_file(&spec, &"rna".into(), None, &"region".into());
        // "file" selector returns union of read and region files
        assert!(all_files.len() >= read_files.len());
        assert!(all_files.len() >= region_files.len());
    }

    #[test]
    fn test_seqspec_file_by_read_id() {
        let spec = dogma_spec();
        let rna_reads = spec.get_seqspec("rna");
        let read_id = rna_reads[0].read_id.clone();
        let ids = vec![read_id.clone()];
        let files = seqspec_file(&spec, &"rna".into(), Some(&ids), &"read".into());
        assert!(files.contains_key(&read_id));
    }

    #[test]
    fn test_full_path() {
        let spec_fn = PathBuf::from("/data/specs/spec.yaml");
        let url = "barcodes.txt".to_string();
        let result = full_path(spec_fn.parent(), &url);
        assert_eq!(result, "/data/specs/barcodes.txt");
    }

    #[test]
    fn test_maybe_full_local() {
        let spec_fn = PathBuf::from("/data/specs/spec.yaml");
        let url = "file.txt".to_string();
        let result = maybe_full(&url, &"local".to_string(), spec_fn.parent(), true);
        assert_eq!(result, "/data/specs/file.txt");
    }

    #[test]
    fn test_maybe_full_remote() {
        let spec_fn = PathBuf::from("/data/specs/spec.yaml");
        let url = "http://example.com/file.txt".to_string();
        let result = maybe_full(&url, &"http".to_string(), spec_fn.parent(), true);
        assert_eq!(result, "http://example.com/file.txt");
    }

    #[test]
    fn test_format_list_files_paired_uses_row_major_order() {
        let mut files = HashMap::new();
        files.insert(
            "rna_R1".to_string(),
            vec![
                File::new(
                    "r1_a".into(),
                    "r1_a.fastq.gz".into(),
                    "fastq".into(),
                    0,
                    "r1_a.fastq.gz".into(),
                    "local".into(),
                    "".into(),
                ),
                File::new(
                    "r1_b".into(),
                    "r1_b.fastq.gz".into(),
                    "fastq".into(),
                    0,
                    "r1_b.fastq.gz".into(),
                    "local".into(),
                    "".into(),
                ),
            ],
        );
        files.insert(
            "rna_R2".to_string(),
            vec![
                File::new(
                    "r2_a".into(),
                    "r2_a.fastq.gz".into(),
                    "fastq".into(),
                    0,
                    "r2_a.fastq.gz".into(),
                    "local".into(),
                    "".into(),
                ),
                File::new(
                    "r2_b".into(),
                    "r2_b.fastq.gz".into(),
                    "fastq".into(),
                    0,
                    "r2_b.fastq.gz".into(),
                    "local".into(),
                    "".into(),
                ),
            ],
        );

        let rendered = format_list_files(
            &files,
            &"paired".to_string(),
            Some(&"filename".to_string()),
            Some(Path::new(".")),
            false,
        );
        assert_eq!(
            rendered,
            "r1_a.fastq.gz\tr2_a.fastq.gz\nr1_b.fastq.gz\tr2_b.fastq.gz"
        );
    }

    #[test]
    fn test_format_list_files_metadata_keeps_raw_local_url() {
        let mut files = HashMap::new();
        files.insert(
            "rna_R1".to_string(),
            vec![File::new(
                "r1".into(),
                "r1.fastq.gz".into(),
                "fastq".into(),
                0,
                "relative/r1.fastq.gz".into(),
                "local".into(),
                "".into(),
            )],
        );

        let rendered = format_list_files_metadata(
            &files,
            &"url".to_string(),
            Some(Path::new("/tmp")),
            true,
        );
        assert_eq!(rendered, "rna_R1\tr1\trelative/r1.fastq.gz");
    }
}
